Generate Tracking event docs from JSDoc

tomalec opened the pull request:
:warning:  This PR requires https://github.com/woocommerce/pinterest-for-woocommerce/pull/237 to have something to generate the docs for.
<!-- You can erase any parts of this template not applicable to your pull request. -->

<!-- Reference issue(s) that this PR fixes or addresses -->
Related to
- https://github.com/woocommerce/pinterest-for-woocommerce/issues/232
- https://github.com/woocommerce/google-listings-and-ads/issues/117

### Changes proposed in this pull request
<!-- Explain what this PR adds or changes, why, and how this will impact users. -->
- Install `jsdoc`, with `jsdoc-plugin-typescript`, 
	to run on our imported types.
- Add JSDoc template for tracking events .md, 
	Add a plugin to resolve `.~` alias.



#### Screenshots
<!--- Optional --->


![jsdoc tracking](https://user-images.githubusercontent.com/17435/140085218-8c65d444-c65f-4355-b9a6-de862bd856e9.gif)

### Detailed test instructions
<!-- Add steps to confirm the fix or change. -->
0. Add/change some JSDoc for and `@event`
1. ```
   npm install
   npm run doc:tracking
   code TRACKING.md
   ```
<!-- Please add details of other areas that might be impacted by the change. -->

### Additional notes:
1. With the upcoming JSDoc release, we can move `.jsdocrc.json` content to `package.json` https://github.com/jsdoc/jsdoc/issues/1550. But there is no easy way to install `jsdoc@4.0.0-dev.17`.
2. Currently the script considers every `@event` a tracking one. In the long term, we'd rather filter them. We may do it later if we start to document other events, or decide now on the means to filter them:
	1. Name convention, and document only one which name starts with `wcadmin_pfw_` - easiest to implement
	1. Add a specific JSDoc tag, like `@tracking` - that would require adapting linter rules
	2. Filter it by `type` - that would require creating that type, deciding where to store `typedef`, then adapt the script to track all imported types down to root one. That may be a bit tricky, given the `{import('foo').bar}` is already sketchy.
3. Eventually, we may separate the JSDoc plugin code to a separate repo, so it could be shared across other Woo extensions.

### Changelog note
<!-- Changelog notes highlight what's fixed or added in a release. -->

> Dev: Generate Tracking event docs from JSDoc

tomalec committed:
Add JSDoc template for trackign events .md,

Add a plugin to resolve .~ alias.

Related to
- https://github.com/woocommerce/pinterest-for-woocommerce/issues/232
- https://github.com/woocommerce/google-listings-and-ads/issues/117

tomalec committed:
Install JSDoc, with `jsdoc-plugin-typescript`,

to run on our imported types.

Related to
- https://github.com/woocommerce/pinterest-for-woocommerce/issues/232
- https://github.com/woocommerce/google-listings-and-ads/issues/117

eason9487 commented on woo-tracking-jsdoc/README.md:
```diff
@@ -0,0 +1,65 @@
+		"path": "TRACKING.md",
+		// Pattern to be used to match the content to be replaced. The groups are respectively: start marker, replaceable content, end marker.
+		"replacement": "(<woo-tracking-jsdoc.*>)([\s\S]*)(<\/woo-tracking-jsdoc.*>)"
+	}
```
💅 
Mixed tab and space indentations.

tomalec replied:
Addressed in https://github.com/woocommerce/pinterest-for-woocommerce/pull/238/commits/198eb37f633e734596fceeed14a04f8bbde0b0bc

eason9487 commented on woo-tracking-jsdoc/README.md:
```diff
@@ -0,0 +1,65 @@
+		"jsdoc-plugin-typescript"
+    ],
+    "typescript":
+	// …
```
💅 
Mixed tab and space indentations.

tomalec replied:
Addressed in https://github.com/woocommerce/pinterest-for-woocommerce/pull/238/commits/198eb37f633e734596fceeed14a04f8bbde0b0bc

eason9487 commented on woo-tracking-jsdoc/README.md:
```diff
@@ -0,0 +1,65 @@
+    ],
+    "typescript": {
+      "moduleRoot": "assets/source" // Path to your module's root directory.
+    }
```
💅 
Inconsistent width unit of indentation.

tomalec replied:
Addressed in https://github.com/woocommerce/pinterest-for-woocommerce/pull/238/commits/198eb37f633e734596fceeed14a04f8bbde0b0bc

eason9487 commented on woo-tracking-jsdoc/README.md:
````diff
@@ -0,0 +1,65 @@
+```json
+{
+    "plugins": [
+      "woo-tracking-jsdoc/skip-import-type.js",
````
❔ 
Is there a description or file about `skip-import-type.js` that is missing?

tomalec replied:
It was a leftover from experiments before finding `jsdoc-plugin-typescript`.
Removed in https://github.com/woocommerce/pinterest-for-woocommerce/pull/238/commits/198eb37f633e734596fceeed14a04f8bbde0b0bc

eason9487 commented on woo-tracking-jsdoc/README.md:
````diff
@@ -0,0 +1,65 @@
+## Config
+
+You may add any of the following properties to your JSDoc config (`.jsdocrc.json`) to change those default values:
+```js
````
Looks like it should be `json`?

```suggestion
```json
```

tomalec replied:
json does not allow comments, but js does.
![image](https://user-images.githubusercontent.com/17435/140314256-6f266c64-c144-4866-9e37-84092d6f48dc.png)

eason9487 commented on .jsdocrc.json:
```diff
@@ -0,0 +1,12 @@
+    ],
+    "typescript": {
+      "moduleRoot": "assets/source"
+    }
```
💅 
Inconsistent width unit of indentation.

tomalec replied:
Addressed in https://github.com/woocommerce/pinterest-for-woocommerce/pull/238/commits/198eb37f633e734596fceeed14a04f8bbde0b0bc

eason9487 commented on woo-tracking-jsdoc/README.md:
```diff
@@ -0,0 +1,65 @@
+		// Path to the markdown file to which the tracking events' docs should be added
+		"path": "TRACKING.md",
+		// Pattern to be used to match the content to be replaced. The groups are respectively: start marker, replaceable content, end marker.
+		"replacement": "(<woo-tracking-jsdoc.*>)([\s\S]*)(<\/woo-tracking-jsdoc.*>)"
```
🚧 

Need to escape `\` in JSON string.

```suggestion
		"replacement": "(<woo-tracking-jsdoc.*>)([\\s\\S]*)(<\\/woo-tracking-jsdoc.*>)"
```

eason9487 commented on woo-tracking-jsdoc/publish.js:
```diff
@@ -0,0 +1,77 @@
+	const {
+		conf: {
+			templates: {
+				'woo-tracking-jsdoc': {
+					path: readmePath = 'TRACKING.md',
+					replacement,
+				} = {},
```
🚧 
The `woo-tracking-jsdoc` layer is inconsistent with its description in *./woo-tracking-jsdoc/README.md*

tomalec replied:
Fixed in https://github.com/woocommerce/pinterest-for-woocommerce/pull/238/commits/17f1c3b7038a2e2eebcdb02b4aab8efabe40dc5a

eason9487 commented on woo-tracking-jsdoc/publish.js:
```diff
@@ -0,0 +1,77 @@
+const defaultReplacementRegex = /(<woo-tracking-jsdoc.*>)([\s\S]*)(<\/woo-tracking-jsdoc.*>)/;
+
+const disclaimer = `<!---
+Everything inside \`<woo-tracking-jsdoc>\` will be automatically generated by \`woo-tracking-jsdoc\`.
```
💅 
How about filling the matched tag here?

```suggestion
Everything inside \`$1\` will be automatically generated by \`woo-tracking-jsdoc\`.
```

Another idea is to add comments at the beginning and end of the auto-generated blocks, but without the tag name. So that when multiple lines of tag and its attributes are matched, there would be no formatting problems when multiple lines of content are placed inside the \`\`.

tomalec replied:
> Another idea is to add comments at the beginning and end of the auto-generated blocks, but without the tag name.

Great idea thanks. Implemented in https://github.com/woocommerce/pinterest-for-woocommerce/pull/238/commits/252faa6a820a9d9b22d3cda868ce03cacb791f25

eason9487 commented on woo-tracking-jsdoc/publish.js:
```diff
@@ -0,0 +1,77 @@
+const path = require( 'jsdoc/path' );
+
+// RegExp used to match the replacement. The groups are respectively: start marker, replacable content, end marker.
+const defaultReplacementRegex = /(<woo-tracking-jsdoc.*>)([\s\S]*)(<\/woo-tracking-jsdoc.*>)/;
```
💅 
Maybe adjust it to `/(<woo-tracking-jsdoc(?:\s[^>]*)?)>([\s\S]*)(<\/woo-tracking-jsdoc>)/`. Even though it's still not precise enough, it could be a proper alternative to handle the below formats if the `.*` was intending to match tag attributes.

```html
Case 1
<woo-tracking-jsdoc foo="1" bar="2">
</woo-tracking-jsdoc>

Case 2
<woo-tracking-jsdoc
  foo="1"
  bar="2"
>
</woo-tracking-jsdoc>
```

But if it intends to match the tag name only, please ignore me. 🙈

tomalec replied:
Great catch, done in https://github.com/woocommerce/pinterest-for-woocommerce/pull/238/commits/0f56bc31eaf769534879c3b8d4848052e9275ddd

eason9487 commented on woo-tracking-jsdoc/tilde-alias.js:
```diff
@@ -0,0 +1,5 @@
+exports.handlers = {
+	beforeParse( e ) {
+		e.source = e.source.replace( /(import\(["'])\.?~\//g, `$1` );
```
❔ 
I'd like to confirm that after replacing a matching content:
```js
/**
 * @property {import(".~/components/app-button").default} helpButton Help button.
*/
```

to
```js
/**
 * @property {import("components/app-button").default} helpButton Help button.
*/
```

then it will be transformed to the below in the *TRACKING.md*

#### Properties
|   |   |   |
|---|---|---|
`helpButton` | `module:components/app-button` | Help button.

Is `module:components/app-button` the expected result?

tomalec replied:
It is expected. 

So far it's pretty primitive as it hardcodes `~` to be the same as `moduleRoot`. I didn't find an easy way to make jsdoc resolve webpack aliases and paths. Hopefully, that's something we will be able to solve in the future. I believe it's along the lines of jsdoc-typescript conventions-webpack

eason9487 commented on woo-tracking-jsdoc/publish.js:
```diff
@@ -0,0 +1,77 @@
+
+	let mdResult = '\n';
+
+	data( { kind: 'event' } ).each( ( symbol ) => {
```
💅 
Maybe add a new line to separate each tracking event block?

![image](https://user-images.githubusercontent.com/17420811/140302954-7e1c02b3-63c8-4e9e-ba38-9eef12f58af7.png)

tomalec replied:
Addressed in https://github.com/woocommerce/pinterest-for-woocommerce/pull/238/commits/59194d8d1cd173e288bffbdd5f5015bddf6d4ac6

tomalec committed:
Fix indentation,

remove not used plugin from example config,
use JS syntax for highlighting JSON with comments in `woo-tracking-jsdoc` files.

Addresses https://github.com/woocommerce/pinterest-for-woocommerce/pull/238#pullrequestreview-797483220

tomalec committed:
Escape `\` in JSON example in `woo-tracking-jsdoc/README.md`

Co-authored-by: Eason Su <eason.su.tw@gmail.com>

Addresses https://github.com/woocommerce/pinterest-for-woocommerce/pull/238#discussion_r742655507

tomalec committed:
Add the `woo-tracking-jsdoc` layer to the `templates` config in `woo-tracking-jsdoc/README.md`

Addresses https://github.com/woocommerce/pinterest-for-woocommerce/pull/238#discussion_r742670060

tomalec committed:
Make the `woo-tracking-jsdoc` disclaimer agnostic to its `replacement`.

Addresses https://github.com/woocommerce/pinterest-for-woocommerce/pull/238#discussion_r742688638

tomalec committed:
Make `woo-tracking-jsdoc`'s replacement match attributes.

Addresses https://github.com/woocommerce/pinterest-for-woocommerce/pull/238#discussion_r742694967

Co-authored-by: Eason Su <eason.su.tw@gmail.com>

tomalec committed:
Add a new line between every Trackign Event docs.

Addresses https://github.com/woocommerce/pinterest-for-woocommerce/pull/238#discussion_r742733381.

tomalec commented:
Thanks, @eason9487 for the detailed review and great ideas. I believe I addressed all your comments. Would you mind another round of review?

eason9487 commented on woo-tracking-jsdoc/README.md:
```diff
@@ -0,0 +1,66 @@
+      // Path to the markdown file to which the tracking events' docs should be added
+      "path": "TRACKING.md",
+      // Pattern to be used to match the content to be replaced. The groups are respectively: start marker, replaceable content, end marker.
+      "replacement": "(<woo-tracking-jsdoc(?:\s[^>]*)?>)([\\s\\S]*)(<\\/woo-tracking-jsdoc.*>)"
```
Escape char typo.

```suggestion
      "replacement": "(<woo-tracking-jsdoc(?:\\s[^>]*)?>)([\\s\\S]*)(<\\/woo-tracking-jsdoc.*>)"
```

eason9487 reviewed (approved):
Thanks for adding this convenient tool to generate tracking event docs! Testing well, and LGTM. Commented on a typo, but not a blocker.

tomalec committed:
Escape `\` in `woo-tracking-jsdoc/README.md`

Co-authored-by: Eason <eason.su.tw@gmail.com>

tomalec committed:
Merge remote-tracking branch 'origin/feature/232-tracking' into add/232-jsdoc-tracks

tomalec committed:
Generate tracking docs for the new events,

added in the merged `feature/232-tracking` branch.
